mmverify.py

```diff
@@ -1,369 +1,588 @@
+
+    def add_d(self, varlist: list[Var]) -> None:
+        """Add a disjoint variable condition (ordered pair of variables) to
+        the frame stack top.
+        """
+        self[-1].d.update((min(x, y), max(x, y))
+                          for x, y in itertools.product(varlist, varlist) if x != y)
+
+    def lookup_c(self, tok: Const) -> bool:
+        """Return whether the given token is an active constant."""
+        return any(tok in fr.c for fr in self)
+
+    def lookup_v(self, tok: Var) -> bool:
+        """Return whether the given token is an active variable."""
+        return any(tok in fr.v for fr in self)
+
+    def lookup_d(self, x: Var, y: Var) -> bool:
+        """Return whether the given ordered pair of tokens belongs to an
+        active disjoint variable statement.
+        """
+        return any((min(x, y), max(x, y)) in fr.d for fr in self)
+
+    def lookup_f(self, var: Var) -> Label:
+        """Return the label of the active floating hypothesis which types the
+        given variable.
+        """
+        for frame in self:
+            try:
+                return frame.f_labels[var]
+            except KeyError:
+                pass
         raise MMKeyError(var)
 
-    def lookup_d(self, x, y):
-        return any(((min(x,y), max(x,y)) in fr.d for fr in reversed(self)))
-
-    def lookup_e(self, stmt):
+    def lookup_e(self, stmt: Stmt) -> Label:
+        """Return the label of the (earliest) active essential hypothesis with
+        the given statement.
+        """
         stmt_t = tuple(stmt)
-        for frame in reversed(self):
-            try: return frame.e_labels[stmt_t]
-            except KeyError: pass
+        for frame in self:
+            try:
+                return frame.e_labels[stmt_t]
+            except KeyError:
+                pass
         raise MMKeyError(stmt_t)
 
-    def make_assertion(self, stat):
-        frame = self[-1]
+    def find_vars(self, stmt: Stmt) -> list[Var]:
+        """Return the list of variables in the given statement."""
+        var_list = []
+        for x in stmt:
+            if x not in var_list and self.lookup_v(x):
+                var_list.append(x)
+        return var_list
+
+    def make_assertion(self, stmt: Stmt) -> Assertion:
+        """Return a quadruple (disjoint variable conditions, floating
+        hypotheses, essential hypotheses, conclusion) describing the given
+        assertion.
+        """
         e_hyps = [eh for fr in self for eh in fr.e]
-        mand_vars = {tok for hyp in itertools.chain(e_hyps, [stat])
-                         for tok in hyp if self.lookup_v(tok)}
-
-        dvs = {(x,y) for fr in self for (x,y) in
+        mand_vars = {tok for hyp in itertools.chain(e_hyps, [stmt])
+                     for tok in hyp if self.lookup_v(tok)}
+        dvs = {(x, y) for fr in self for (x, y) in
                fr.d.intersection(itertools.product(mand_vars, mand_vars))}
```

Comment:
Once again not your code, but [this SO post](https://stackoverflow.com/questions/6821329/speed-differences-between-intersection-and-object-for-object-in-set-if-object) suggests `(x, y) in fr.d and x in mand_vars and y in mand_vars` would be more performant than `intersection(product(..))`.

Reply:
I'm not fluent in Python set comprehension, but I guess you meant `dvs = {(x, y) for fr in self for (x, y) in fr.d if x in mand_vars and y in mand_vars}`. Implemented in next commit.